### ISSUE TITLE
Allow for different brackets/braces spacing

### ISSUE THREAD

**Issue by langdonx** [eslint/eslint#1576]
The `space-in-brackets` rule forces spaces surrounding braces/brackets to be uniform, although I have always preferred space-in-**braces**, but not space-in-**brackets**.

I do my arrays like this:

    var a = [1, 2, 3];
        b = ['1', '2', '3'];

And my objects like this:

    var a = { a: 1, b: 2, c: 3 }
        b = { a: [1, 2, 3] },
        c = { a: [1, 2, 3], b: true }

**Comment by vitorbal** [comment]
Agreed, in my current project we would like to enforce space in brackets, but no space in braces. Would be great if this was possible!

**Comment by kentor** [comment]
:+1: yeah I use this convention too

**Comment by nzakas** [comment]
This issue is open for implementation, patches welcome! :)

**Comment by JedWatson** [comment]
I follow the same conventions as @langdonx, and I'm happy to have a go at submitting a patch to fix it :)

Looks like it the fix be as simple as expanding `options.spaced` into `options.arraySpaced` and `options.objectSpaced` then checking those more specific options in the `ArrayExpression` and `ObjectExpression` methods.

Only tricky thing is making sure the logic flips are correct, I'll add tests to make sure it all works correctly.

Let me know if you think that approach isn't suitable.

**Comment by nzakas** [comment]
I think the better approach is to split into two rules: `space-in-brackets` and `space-in-braces`

**Comment by JedWatson** [comment]
@nzakas I fixed up the commit message on that PR.

I agree this would make more sense as two rules. I figured given the history in this issue that wasn't the goal however so I implemented it as per the discussion above.

Given that the fix actually works, has documentation and tests, perhaps you could merge this into the next release and we could work on replacing it with two rules later (new issue for 1.0.0 maybe?) - there's precedent with other rules becoming deprecated.

I worked on this because I'm implementing ESLint across several large projects and want to be able to enable this rule, but cannot until these options (or equivalent) are supported.

**Comment by nzakas** [comment]
I'd prefer not to introduce two sets of changes for this. It's very hard to make breaking changes, and given we are so close to v1.0.0, I'd like the fix for this to be the final one.

So that means either what you worked on is the final version or the split rules is the final version. 

My gut feeling is that if we go with your version, we will end up seeing a bug related to the rule naming pop up again. I admit I was unsure of the right approach and could have been more clear in this issue about what I was thinking, so I apologize for that. Now I think having this split into two rules makes the most sense.

If you are blocked on this, you should feel free to take your work and bundle it as a plugin that you can use.

**Comment by andrewaramsay** [comment]
Just a clarification.  I believe **brackets** are the term for the characters used for array literals [], and **braces** are the term for object literals {}.  In the original post it seems they are backwards.

[See Wikipedia](http://en.wikipedia.org/wiki/Bracket#Names_for_various_bracket_symbols)

Unless I am unaware of some JavaScript convention which uses the opposite names.

**Comment by langdonx** [comment]
@andrewaramsay Interesting, and good call.  I've amended my original post in light of this.

To make it perfectly clear, the rules could be named `space-in-curly-braces` and `space-in-brackets`.

**Comment by JedWatson** [comment]
@nzakas,

I'm on board with one change set for this and that two rules are clearer going forward... and I'm happy to have a go at implementing the new new rules to replace this one.

I've opened two new issues to track the new rules so they can have clearer individual discussions and PRs. If you're happy with this approach please add the "accepted" tag and I'll start work on them.

It's up to you if you'd like to provide direction on which options they should support from the set currently implemented in `spaces-in-brackets`... otherwise I'm happy to have a look at what makes sense. Personally I find the current options implementation a bit obscure and this might be a good opportunity to simplify them...

@langdonx it's probably best to pick entirely new names (not re-use `space-in-brackets`) so the current behaviour can be deprecated, rather than just changing it. Otherwise users not following so closely could easily miss that it's now just applied to Array notation.

Also since understanding of the different names (brackets, braces, etc) is not universal, maybe using terms like "array notation" and "object notation" would be clearer.

**Comment by bytasv** [comment]
@JedWatson I agree on using new names for separate rules and old rule could just simply throw a deprecation warning when used in newer versions...

**Comment by justin808** [comment]
Any ETA for this issue? This is breaking conformance with https://github.com/airbnb/javascript/tree/es6#whitespace

where there's [a, b] and { a:1, b:2 } in the examples

**Comment by nzakas** [comment]
This was split into #2225 and #2226.
